File: sorting_algos/merge_sort.c
#include <stdio.h>
#include <stdlib.h>

void inline_merge_sort(int *array, int start, int end){
    int *left, *right;
    if(end-start<=1){
        return;
    }
    printf("%d %d ", start, end);
    int mid = (start+end)/2;
    left = (int *)malloc((mid-start)*sizeof(int));
    right = (int *)malloc((end-mid)*sizeof(int));
    for(int i=start; i<mid; i++){
        left[i] = array[i];
        printf("left %d\n",array[i]);
    }
    for(int i=mid;i<end; i++){
        right[i-mid] = array[i];
        printf("right %d\n",array[i]);
    }
    inline_merge_sort(left, start, mid);
    inline_merge_sort(right, 0, end-mid);
    int i=0,j=0,k=0,imax=mid,jmax=end-mid;
    while (i<imax && j<jmax)
    {
        if(left[i]<right[j]) array[k]=left[i++];
        else array[k]=right[j++];
        k++;
    }
    while(i<imax){
        array[k++]=left[i++];
    }
    while(j<jmax){
        array[k++]=right[j++];
    }
}

int main(){
    int arr[]={3,4,2,1,5}, len=5;
    inline_merge_sort(arr,0,len);
    for(int i=0;i<len;i++){
        printf("%d\t",arr[i]);
    }
    printf("\n");
    return 0;
}

File: sorting_algos/merge_sort.rs
use std::fmt::{Display,Debug};

use rand::Rng;

fn merge_sort<T>(v: Vec<T>) ->  Vec<T> where T:Clone+PartialOrd+Display+Debug{
    if v.len()<=1{return v}
    let mid=v.len()/2;
    let mut newv = Vec::with_capacity(v.len());
    newv.resize(v.len(), v[0].clone());
    let (left,right)=v.split_at(mid);
    let sortedleft = merge_sort(left.to_vec());
    let sortedright = merge_sort(right.to_vec());
    let [mut i, mut j, mut k]=[0,0,0];
    let [imax,jmax]=[sortedleft.len(),sortedright.len()];
    while i<imax && j<jmax{
        if sortedleft[i]<sortedright[j]{
            newv[k] = sortedleft[i].clone();
            i+=1;
        }
        else{
            newv[k] = sortedright[j].clone();
            j+=1;
        }
        k+=1;
    }
    while i<imax{
        newv[k] = sortedleft[i].clone();
        i+=1;
        k+=1;
    }
    while j<jmax{
        newv[k] = sortedright[j].clone();
        j+=1;
        k+=1;
    }
    newv
}
fn main(){
    let rawv = [0; 10];
    let mut rng=rand::thread_rng();
    let v = rawv.map(|_|{rng.gen_range(0..100)}).to_vec();
    println!("{:?}",v);
    let v2 = merge_sort(v);
    println!("{:?}",v2);
}